File: 309.最佳买卖股票时机含冷冻期.java
/*
 * @lc app=leetcode.cn id=309 lang=java
 *
 * [309] 最佳买卖股票时机含冷冻期
 */

// @lc code=start
class Solution {
    public int maxProfit(int[] prices) {
        int len = prices.length;
        // 特殊判断
        if (len < 2) {
            return 0;
        }
        int[][] dp = new int[len][2];

        // dp[i][0] 下标为 i 这天结束的时候，不持股，手上拥有的现金数
        // dp[i][1] 下标为 i 这天结束的时候，持股，手上拥有的现金数

        // 初始化：不持股显然为 0，持股就需要减去第 1 天（下标为 0）的股价
        dp[0][0] = 0;
        dp[0][1] = -prices[0];

        // 从第 2 天开始遍历
        for (int i = 1; i < len; i++) {
            if (i == 1) {
                dp[i][1] = Math.max(dp[i - 1][1], -prices[i]);
            } else {
                dp[i][1] = Math.max(dp[i - 1][1], dp[i - 2][0] - prices[i]);
            }
            dp[i][0] = Math.max(dp[i - 1][0], dp[i - 1][1] + prices[i]);

        }
        return dp[len - 1][0];
    }
}
// @lc code=end

File: 309.最佳买卖股票时机含冷冻期.cpp
/*
 * @lc app=leetcode.cn id=309 lang=cpp
 *
 * [309] 最佳买卖股票时机含冷冻期
 */

// @lc code=start
class Solution
{
public:
    int maxProfit(vector<int> &prices)
    {
        int dp_0 = 0;
        int dp_1 = 0;
        int dp_pre_0 = 0;
        for (int i = 0; i < prices.size(); i++)
        {
            if (i == 0)
            {
                dp_0 = 0;
                dp_1 = -prices[i];
            }
            else
            {
                int a = dp_0;
                dp_0 = max(dp_0, dp_1 + prices[i]);
                dp_1 = max(dp_1, dp_pre_0 - prices[i]);
                dp_pre_0 = a;
            }
        }
        return dp_0;
    }
};
// @lc code=end
